# Prediction of Chronic Kidney Disease

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


# Data preprocessing 

In [2]:
import numpy as np

In [33]:
import pandas as pd
#importing the dataset into pandas dataframe 
df = pd.read_excel(r"/content/drive/My Drive/chronic_kidney_disease.xlsx")
df

age   bp     sg   al   su     rbc  ...   dm  cad appet   pe  ane   class
0     48   80  1.020    1    0       ?  ...  yes   no  good   no   no     ckd
1      7   50  1.020    4    0       ?  ...   no   no  good   no   no     ckd
2     62   80  1.010    2    3  normal  ...  yes   no  poor   no  yes     ckd
3     48   70  1.005    4    0  normal  ...   no   no  poor  yes  yes     ckd
4     51   80  1.010    2    0  normal  ...   no   no  good   no   no     ckd
..   ...  ...    ...  ...  ...     ...  ...  ...  ...   ...  ...  ...     ...
396   42   70  1.025    0    0  normal  ...   no   no  good   no   no  notckd
397   12   80  1.020    0    0  normal  ...   no   no  good   no   no  notckd
398   17   60  1.025    0    0  normal  ...   no   no  good   no   no  notckd
399   58   80  1.025    0    0  normal  ...   no   no  good   no   no  notckd
400  NaN  NaN    NaN  NaN  NaN     NaN  ...  NaN  NaN   NaN  NaN  NaN     NaN

[401 rows x 25 columns]

In [34]:
df.drop(df.tail(1).index,inplace=True)
df

age  bp     sg al su     rbc  ...   dm cad appet   pe  ane   class
0    48  80  1.020  1  0       ?  ...  yes  no  good   no   no     ckd
1     7  50  1.020  4  0       ?  ...   no  no  good   no   no     ckd
2    62  80  1.010  2  3  normal  ...  yes  no  poor   no  yes     ckd
3    48  70  1.005  4  0  normal  ...   no  no  poor  yes  yes     ckd
4    51  80  1.010  2  0  normal  ...   no  no  good   no   no     ckd
..   ..  ..    ... .. ..     ...  ...  ...  ..   ...  ...  ...     ...
395  55  80  1.020  0  0  normal  ...   no  no  good   no   no  notckd
396  42  70  1.025  0  0  normal  ...   no  no  good   no   no  notckd
397  12  80  1.020  0  0  normal  ...   no  no  good   no   no  notckd
398  17  60  1.025  0  0  normal  ...   no  no  good   no   no  notckd
399  58  80  1.025  0  0  normal  ...   no  no  good   no   no  notckd

[400 rows x 25 columns]

In [35]:
df.replace('?',np.nan,inplace=True)
df.replace('\t?',np.nan,inplace=True)

In [36]:
df.replace('\tno','no',inplace=True)
df.replace('\tyes','yes',inplace=True)
df.replace(' yes','yes',inplace=True)
df.replace('ckd\t','ckd',inplace=True)

In [32]:
df.replace('yes',1,inplace=True)
df.replace('no',0,inplace=True)
df.replace('normal',0,inplace=True)
df.replace('abnormal',1,inplace=True)
df.replace('notpresent',0,inplace=True)
df.replace('present',1,inplace=True)
df.replace('poor',0,inplace=True)
df.replace('good',1,inplace=True)
df.replace('notckd',0,inplace=True)
df.replace('ckd',1,inplace=True)

In [37]:
#df=pd.get_dummies(df)
df.iloc[[294]]

age  bp     sg al su     rbc      pc  ... htn  dm cad appet   pe  ane   class
294  75  60  1.020  0  0  normal  normal  ...  no  no  no   NaN  NaN  NaN  notckd

[1 rows x 25 columns]

In [38]:
cat_columns=['bp','sg','al','su','rbc','pc','pcc','ba','htn','dm','cad','appet','pe','ane','class']
for x in df.columns:
  print(x)
  print(sum(df[x].isna()))
  #df[x].str.replace("\t"," ")
  #print(df[x].value_counts())

age
9
bp
12
sg
47
al
46
su
49
rbc
152
pc
65
pcc
4
ba
4
bgr
44
bu
19
sc
17
sod
87
pot
88
hemo
52
pcv
71
wbcc
106
rbcc
131
htn
2
dm
2
cad
2
appet
1
pe
1
ane
1
class
0


In [16]:
#df.ane.isna()
df[df['appet'].isnull()].index.tolist()

[294]

In [39]:
#Finding mean,minimum and maximum values to find the ouliers
df.describe()

age   bp     sg   al   su     rbc  ...   dm  cad appet   pe  ane class
count   391  388    353  354  351     248  ...  398  398   399  399  399   400
unique   77   12      5    6    6       2  ...    2    2     2    2    2     2
top      60   80  1.020    0    0  normal  ...   no   no  good   no   no   ckd
freq     19  115    106  199  290     201  ...  261  364   317  323  339   250

[4 rows x 25 columns]

In [40]:
df_required = df.drop(['rbc','wbcc','rbcc'], axis=1)
df_required

age  bp     sg al su        pc  ...   dm cad appet   pe  ane   class
0    48  80  1.020  1  0    normal  ...  yes  no  good   no   no     ckd
1     7  50  1.020  4  0    normal  ...   no  no  good   no   no     ckd
2    62  80  1.010  2  3    normal  ...  yes  no  poor   no  yes     ckd
3    48  70  1.005  4  0  abnormal  ...   no  no  poor  yes  yes     ckd
4    51  80  1.010  2  0    normal  ...   no  no  good   no   no     ckd
..   ..  ..    ... .. ..       ...  ...  ...  ..   ...  ...  ...     ...
395  55  80  1.020  0  0    normal  ...   no  no  good   no   no  notckd
396  42  70  1.025  0  0    normal  ...   no  no  good   no   no  notckd
397  12  80  1.020  0  0    normal  ...   no  no  good   no   no  notckd
398  17  60  1.025  0  0    normal  ...   no  no  good   no   no  notckd
399  58  80  1.025  0  0    normal  ...   no  no  good   no   no  notckd

[400 rows x 22 columns]

In [60]:
grouped = df_required.groupby(df['class'])
ckd = grouped.get_group("ckd")
notckd = grouped.get_group("notckd")

In [61]:
ckd

age   bp     sg   al   su        pc  ...   dm  cad appet   pe  ane class
0    48   80  1.020    1    0    normal  ...  yes   no  good   no   no   ckd
1     7   50  1.020    4    0    normal  ...   no   no  good   no   no   ckd
2    62   80  1.010    2    3    normal  ...  yes   no  poor   no  yes   ckd
3    48   70  1.005    4    0  abnormal  ...   no   no  poor  yes  yes   ckd
4    51   80  1.010    2    0    normal  ...   no   no  good   no   no   ckd
..   ..  ...    ...  ...  ...       ...  ...  ...  ...   ...  ...  ...   ...
245  48  100    NaN  NaN  NaN       NaN  ...   no  yes  poor   no   no   ckd
246  48  110  1.015    3    0    normal  ...   no  yes  good   no  yes   ckd
247  54   90  1.025    1    0  abnormal  ...   no   no  poor  yes  yes   ckd
248  59   70  1.010    1    3  abnormal  ...  yes  yes  good   no   no   ckd
249  56   90  1.010    4    1  abnormal  ...  yes   no  poor  yes  yes   ckd

[250 rows x 22 columns]

In [86]:
#ckd.convert_dtypes()
ckd.dtypes

age        int64
bp        object
sg        object
al        object
su        object
pc        object
pcc       object
ba        object
bgr        int64
bu       float64
sc       float64
sod      float64
pot      float64
hemo     float64
pcv        int64
htn       object
dm        object
cad       object
appet     object
pe        object
ane       object
class     object
dtype: object

In [73]:
ckd['age'] = ckd['age'].fillna(0)
ckd['bgr'] = ckd['bgr'].fillna(0)
ckd['bu'] = ckd['bu'].fillna(0)
ckd['sc'] = ckd['sc'].fillna(0)
ckd['sod'] = ckd['sod'].fillna(0)
ckd['pot'] = ckd['pot'].fillna(0)
ckd['hemo'] = ckd['hemo'].fillna(0)
ckd['pcv'] = ckd['pcv'].fillna(0)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

In [83]:
ckd['age']=ckd['age'].astype('int64')
ckd['bgr']=ckd['bgr'].astype('int64')
ckd['bu']=ckd['bu'].astype('float64')
ckd['sc']=ckd['sc'].astype('float64')
ckd['sod']=ckd['sod'].astype('float64')
ckd['pot']=ckd['pot'].astype('float64')
ckd['hemo']=ckd['hemo'].astype('float64')
ckd['pcv']=ckd['pcv'].astype('int64')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [89]:
ckd['age'] = ckd['age'].fillna((ckd['age'].mean()))
ckd['bgr'] = ckd['bgr'].fillna((ckd['bgr'].mean()))
ckd['bu'] = ckd['bu'].fillna((ckd['bu'].mean()))
ckd['sc'] = ckd['sc'].fillna((ckd['sc'].mean()))
ckd['sod'] = ckd['sod'].fillna((ckd['sod'].mean()))
ckd['pot'] = ckd['pot'].fillna((ckd['pot'].mean()))
ckd['hemo'] = ckd['hemo'].fillna((ckd['hemo'].mean()))
ckd['pcv'] = ckd['pcv'].fillna((ckd['pcv'].mean()))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

In [95]:
#ckd.loc[ckd['bgr'] == 0, 'bgr' ] = ckd['bgr'].mean()
ckd.loc[ckd['bu'] == 0, 'bu' ] = ckd['bu'].mean()
ckd.loc[ckd['sc'] == 0, 'sc' ] = ckd['sc'].mean()
ckd.loc[ckd['sod'] == 0, 'sod' ] = ckd['sod'].mean()
ckd.loc[ckd['pot'] == 0, 'pot' ] = ckd['pot'].mean()
ckd.loc[ckd['hemo'] == 0, 'hemo' ] = ckd['hemo'].mean()
ckd.loc[ckd['pcv'] == 0, 'pcv' ] = ckd['pcv'].mean()
ckd.loc[ckd['age'] == 0, 'age' ] = ckd['age'].mean()

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [100]:
ckd['bp'] = ckd['bp'].astype('category')
ckd['sg'] = ckd['sg'].astype('category')
ckd['al'] = ckd['al'].astype('category')
ckd['su'] = ckd['su'].astype('category')
ckd['pc'] = ckd['pc'].astype('category')
ckd['pcc'] = ckd['pcc'].astype('category')
ckd['ba'] = ckd['ba'].astype('category')
ckd['htn'] = ckd['htn'].astype('category')
ckd['dm'] = ckd['dm'].astype('category')
ckd['cad'] = ckd['cad'].astype('category')
ckd['appet'] = ckd['appet'].astype('category')
ckd['pe'] = ckd['pe'].astype('category')
ckd['ane'] = ckd['ane'].astype('category')
ckd['class'] = ckd['class'].astype('category')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

In [102]:
ckd['bp'] = ckd['bp'].fillna((ckd['bp'].mode()))
ckd['sg'] = ckd['sg'].fillna((ckd['sg'].mode()))
ckd['al'] = ckd['al'].fillna((ckd['al'].mode()))
ckd['su'] = ckd['su'].fillna((ckd['su'].mode()))
ckd['pc'] = ckd['pc'].fillna((ckd['pc'].mode()))
ckd['pcc'] = ckd['pcc'].fillna((ckd['pcc'].mode()))
ckd['ba'] = ckd['ba'].fillna((ckd['ba'].mode()))
ckd['htn'] = ckd['htn'].fillna((ckd['htn'].mode()))
ckd['dm'] = ckd['dm'].fillna((ckd['dm'].mode()))
ckd['cad'] = ckd['cad'].fillna((ckd['cad'].mode()))
ckd['appet'] = ckd['appet'].fillna((ckd['appet'].mode()))
ckd['pe'] = ckd['pe'].fillna((ckd['pe'].mode()))
ckd['ane'] = ckd['ane'].fillna((ckd['ane'].mode()))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

In [ ]:
cat_columns=['bp','sg','al','su','pc','pcc','ba','htn','dm','cad','appet','pe','ane','class']

In [125]:
ckd.dtypes

age      float64 
bp       category
sg       category
al       category
su       category
pc       category
pcc      category
ba       category
bgr      float64 
bu       float64 
sc       float64 
sod      float64 
pot      float64 
hemo     float64 
pcv      float64 
htn      category
dm       category
cad      category
appet    category
pe       category
ane      category
class    category
dtype: object

In [124]:
ckd.isna().sum().sum()

0

In [122]:
#ckd.bp.mode()
#ckd.bp.isna().sum()
navaluecolumns=['bp','sg','al','su','pc']
for column in navaluecolumns:
    ckd[column].fillna(ckd[column].mode()[0], inplace=True)

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [121]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)
ckd

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  after removing the cwd from sys.path.


age   bp     sg   al   su        pc         pcc          ba      bgr  \
0    48.000  80   1.020  1    0    normal    notpresent  notpresent  121.000   
1    7.000   50   1.020  4    0    normal    notpresent  notpresent  148.756   
2    62.000  80   1.010  2    3    normal    notpresent  notpresent  423.000   
3    48.000  70   1.005  4    0    abnormal  present     notpresent  117.000   
4    51.000  80   1.010  2    0    normal    notpresent  notpresent  106.000   
5    60.000  90   1.015  3    0    NaN       notpresent  notpresent  74.000    
6    68.000  70   1.010  0    0    normal    notpresent  notpresent  100.000   
7    24.000  NaN  1.015  2    4    abnormal  notpresent  notpresent  410.000   
8    52.000  100  1.015  3    0    abnormal  present     notpresent  138.000   
9    53.000  90   1.020  2    0    abnormal  present     notpresent  70.000    
10   50.000  60   1.010  2    4    abnormal  present     notpresent  490.000   
11   63.000  70   1.010  3    0    abnormal  present     notpresent  380.000   
12   68.000  70   1.015  3    1    normal    present     notpresent  208.000   
13   68.000  70   NaN    NaN  NaN  NaN       notpresent  notpresent  98.000    
14   68.000  80   1.010  3    2    abnormal  present     present     157.000   
15   40.000  80   1.015  3    0    normal    notpresent  notpresent  76.000    
16   47.000  70   1.015  2    0    normal    notpresent  notpresent  99.000    
17   47.000  80   NaN    NaN  NaN  NaN       notpresent  notpresent  114.000   
18   60.000  100  1.025  0    3    normal    notpresent  notpresent  263.000   
19   62.000  60   1.015  1    0    abnormal  present     notpresent  100.000   
20   61.000  80   1.015  2    0    abnormal  notpresent  notpresent  173.000   
21   60.000  90   NaN    NaN  NaN  NaN       notpresent  notpresent  148.756   
22   48.000  80   1.025  4    0    abnormal  notpresent  notpresent  95.000    
23   21.000  70   1.010  0    0    normal    notpresent  notpresent  148.756   
24   42.000  100  1.015  4    0    abnormal  notpresent  present     148.756   
25   61.000  60   1.025  0    0    normal    notpresent  notpresent  108.000   
26   75.000  80   1.015  0    0    normal    notpresent  notpresent  156.000   
27   69.000  70   1.010  3    4    abnormal  notpresent  notpresent  264.000   
28   75.000  70   NaN    1    3    NaN       notpresent  notpresent  123.000   
29   68.000  70   1.005  1    0    abnormal  present     notpresent  148.756   
30   52.796  70   NaN    NaN  NaN  NaN       notpresent  notpresent  93.000    
31   73.000  90   1.015  3    0    abnormal  present     notpresent  107.000   
32   61.000  90   1.010  1    1    normal    notpresent  notpresent  159.000   
33   60.000  100  1.020  2    0    abnormal  notpresent  notpresent  140.000   
34   70.000  70   1.010  1    0    NaN       present     present     171.000   
35   65.000  90   1.020  2    1    normal    notpresent  notpresent  270.000   
36   76.000  70   1.015  1    0    normal    notpresent  notpresent  92.000    
37   72.000  80   NaN    NaN  NaN  NaN       notpresent  notpresent  137.000   
38   69.000  80   1.020  3    0    normal    notpresent  notpresent  148.756   
39   82.000  80   1.010  2    2    NaN       notpresent  notpresent  140.000   
40   46.000  90   1.010  2    0    abnormal  notpresent  notpresent  99.000    
41   45.000  70   1.010  0    0    normal    notpresent  notpresent  148.756   
42   47.000  100  1.010  0    0    normal    notpresent  notpresent  204.000   
43   35.000  80   1.010  1    0    NaN       notpresent  notpresent  79.000    
44   54.000  80   1.010  3    0    abnormal  notpresent  notpresent  207.000   
45   54.000  80   1.020  3    0    abnormal  notpresent  notpresent  208.000   
46   48.000  70   1.015  0    0    normal    notpresent  notpresent  124.000   
47   11.000  80   1.010  3    0    normal    notpresent  notpresent  148.756   
48   73.000  70   1.005  0    0    normal    notpresent  notpresent  70.000    
49   60.

In [126]:
notckd

age   bp     sg   al   su      pc         pcc          ba  bgr   bu   sc  \
250  40   80   1.025  0    0    normal  notpresent  notpresent  140  10   1.2   
251  23   80   1.025  0    0    normal  notpresent  notpresent  70   36   1.0   
252  45   80   1.025  0    0    normal  notpresent  notpresent  82   49   0.6   
253  57   80   1.025  0    0    normal  notpresent  notpresent  119  17   1.2   
254  51   60   1.025  0    0    normal  notpresent  notpresent  99   38   0.8   
255  34   80   1.025  0    0    normal  notpresent  notpresent  121  27   1.2   
256  60   80   1.025  0    0    normal  notpresent  notpresent  131  10   0.5   
257  38   60   1.020  0    0    normal  notpresent  notpresent  91   36   0.7   
258  42   80   1.020  0    0    normal  notpresent  notpresent  98   20   0.5   
259  35   80   1.020  0    0    normal  notpresent  notpresent  104  31   1.2   
260  30   80   1.020  0    0    normal  notpresent  notpresent  131  38   1.0   
261  49   80   1.020  0    0    normal  notpresent  notpresent  122  32   1.2   
262  55   80   1.020  0    0    normal  notpresent  notpresent  118  18   0.9   
263  45   80   1.020  0    0    normal  notpresent  notpresent  117  46   1.2   
264  42   80   1.020  0    0    normal  notpresent  notpresent  132  24   0.7   
265  50   80   1.020  0    0    normal  notpresent  notpresent  97   40   0.6   
266  55   80   1.020  0    0    normal  notpresent  notpresent  133  17   1.2   
267  48   80   1.025  0    0    normal  notpresent  notpresent  122  33   0.9   
268  NaN  80   NaN    NaN  NaN  NaN     notpresent  notpresent  100  49   1.0   
269  25   80   1.025  0    0    normal  notpresent  notpresent  121  19   1.2   
270  23   80   1.025  0    0    normal  notpresent  notpresent  111  34   1.1   
271  30   80   1.025  0    0    normal  notpresent  notpresent  96   25   0.5   
272  56   80   1.025  0    0    normal  notpresent  notpresent  139  15   1.2   
273  47   80   1.020  0    0    normal  notpresent  notpresent  95   35   0.9   
274  19   80   1.020  0    0    normal  notpresent  notpresent  107  23   0.7   
275  52   80   1.020  0    0    normal  notpresent  notpresent  125  22   1.2   
276  20   60   1.025  0    0    normal  notpresent  notpresent  NaN  NaN  NaN   
277  46   60   1.025  0    0    normal  notpresent  notpresent  123  46   1.0   
278  48   60   1.020  0    0    normal  notpresent  notpresent  112  44   1.2   
279  24   70   1.025  0    0    normal  notpresent  notpresent  140  23   0.6   
280  47   80   NaN    NaN  NaN  NaN     notpresent  notpresent  93   33   0.9   
281  55   80   1.025  0    0    normal  notpresent  notpresent  130  50   1.2   
282  20   70   1.020  0    0    normal  notpresent  notpresent  123  44   1.0   
283  60   70   1.020  0    0    normal  notpresent  notpresent  NaN  NaN  NaN   
284  33   80   1.025  0    0    normal  notpresent  notpresent  100  37   1.2   
285  66   70   1.020  0    0    normal  notpresent  notpresent  94   19   0.7   
286  71   70   1.020  0    0    normal  notpresent  notpresent  81   18   0.8   
287  39   70   1.025  0    0    normal  notpresent  notpresent  124  22   0.6   
288  56   70   1.025  0    0    normal  notpresent  notpresent  70   46   1.2   
289  42   70   1.020  0    0    normal  notpresent  notpresent  93   32   0.9   
290  54   70   1.020  0    0    NaN     NaN         NaN         76   28   0.6   
291  47   80   1.025  0    0    normal  notpresent  notpresent  124  44   1.0   
292  30   80   1.020  0    0    normal  notpresent  notpresent  89   42   0.5   
293  50   NaN  1.020  0    0    normal  notpresent  notpresent  92   19   1.2   
294  75   60   1.020  0    0    normal  notpresent  notpresent  110  50   0.7   
295  44   70   NaN    NaN  NaN  NaN     notpresent  notpresent  106  25   0.9   
296  41   70   1.020  0    0    normal  notpresent  notpresent  125  38   0.6   
297  53   60   1.025  0    0    normal  notpresent  notpresent  116  26   1.0   
298  34   60   1.020  0    0    norm

In [149]:
notckd.isna().sum().sum()
#notckd['class'] = notckd['class'].astype('category')

0

In [132]:
listtofillnawith0=['age','bgr','bu','sc','sod','pot','hemo','pcv']
#for x in listtofillnawith0:
  #notckd[x] = notckd[x].fillna(0)
intlist=['age','bgr','pcv']
floatlist=['bu','sc','sod','pot','hemo']
cat_columns=['bp','sg','al','su','pc','pcc','ba','htn','dm','cad','appet','pe','ane']
for x in intlist:
  notckd[x] = notckd[x].astype('int64')
for y in floatlist:
  notckd[y] = notckd[y].astype('float64')
for z in cat_columns:
  notckd[z] = notckd[z].astype('category')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the d

In [148]:
ckd['bp'] = ckd['bp'].fillna((ckd['bp'].mode()))
ckd.loc[ckd['bu'] == 0, 'bu' ] = ckd['bu'].mean()
listtofillnawith0=['age','bgr','bu','sc','sod','pot','hemo','pcv']
#for x in listtofillnawith0:
  #notckd.loc[notckd[x]== 0,x] = notckd[x].mean()
intlist=['age','bgr','pcv']
floatlist=['bu','sc','sod','pot','hemo']
cat_columns=['bp','sg','al','su','pc','pcc','ba','htn','dm','cad','appet','pe','ane']
#for x in intlist:
  #notckd[x] = notckd[x].astype('int64')
#for y in floatlist:
  #notckd[y] = notckd[y].astype('float64')
for z in cat_columns:
  notckd[z].fillna(notckd[z].mode()[0], inplace=True)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.o

In [150]:
kidney_disease = pd.concat([ckd,notckd], ignore_index=True)

In [169]:
#kidney_disease.dtypes
#cat_columns=['bp','sg','al','su','pc','pcc','ba','htn','dm','cad','appet','pe','ane','class']
#for z in cat_columns:
   #kidney_disease[z] = kidney_disease[z].astype('category')
   #kidney_disease[z] = kidney_disease[z].cat.codes
#kidney_disease

There are no outliers in the data as all the values in the respective columns are close to the mean value of the columns

# Scaling the data

In [172]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
scaler.fit(kidney_disease)
scaler.fit_transform(kidney_disease)

array([[0.52272727, 0.90909091, 0.75      , ..., 0.        , 0.        ,
        0.        ],
       [0.05681818, 0.63636364, 0.75      , ..., 0.        , 0.        ,
        0.        ],
       [0.68181818, 0.90909091, 0.25      , ..., 0.        , 1.        ,
        0.        ],
       ...,
       [0.11363636, 0.90909091, 0.75      , ..., 0.        , 0.        ,
        1.        ],
       [0.17045455, 0.72727273, 1.        , ..., 0.        , 0.        ,
        1.        ],
       [0.63636364, 0.90909091, 1.        , ..., 0.        , 0.        ,
        1.        ]])

Scaled the dataset using MinMaxScaler 

# Decision tree

In [179]:
from sklearn import tree
#converting the pandas dataframe to arrays
X = kidney_disease[kidney_disease.columns[0:21]]
y = kidney_disease["class"]
X=scaler.fit_transform(X)

0

In [183]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()
clf = clf.fit(X,y)

In [184]:
from sklearn.model_selection import train_test_split  
from sklearn.metrics import accuracy_score 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 100) 

In [185]:
y_pred_DecisionTreeClassifier = clf.predict(X_test) 
print ("Desicion Tree Accuracy is ",accuracy_score(y_test, y_pred_DecisionTreeClassifier)*100 )

Desicion Tree Accuracy is  100.0


In [186]:
y_pred_DecisionTreeClassifier

array([0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0], dtype=int8)

In [187]:
#Using 10 fold cross validation
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X, y, cv=10))
Decision_tree_scores=cross_val_score(clf, X, y, cv=10)

[0.975 1.    1.    1.    0.975 1.    1.    0.975 1.    0.975]


In [188]:
import statistics
statistics.mean(Decision_tree_scores)*100

99.0

Accuracy of the Decision tree while using 10 fold cross validation is 99

# Random Forest

In [189]:
from sklearn.ensemble import RandomForestClassifier

In [190]:
randomforest =RandomForestClassifier(n_estimators=10)

In [191]:
randomforest.fit(X,y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [192]:
y_pred_DecisionTreeClassifier = randomforest.predict(X_test) 
print ("Random Forest Accuracy is ",accuracy_score(y_test, y_pred_DecisionTreeClassifier)*100 )

Random Forest Accuracy is  100.0


In [205]:
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X, y, cv=10))
Random_Forest_scores=cross_val_score(clf, X, y, cv=10)
statistics.mean(Random_Forest_scores)*100

[1.    1.    1.    1.    0.95  1.    0.95  0.975 1.    1.   ]


98.5

In [207]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred_DecisionTreeClassifier)

array([[64,  0],
       [ 0, 36]])

Accuracy of the Random Forest using 10 fold cross validation is 98.75.Accuracy of the Random Forest is higher than Decision tree

# KNN classifier

In [194]:
from sklearn.neighbors import KNeighborsClassifier

In [196]:
knnclassifier =KNeighborsClassifier(n_neighbors=10)

In [197]:
knn_classifier_scores=cross_val_score(knnclassifier, X, y, cv=10)
statistics.mean(knn_classifier_scores)

0.975

In [198]:
knnclassifier =KNeighborsClassifier()

In [199]:
for k in range(1,50):
    knnclassifier =KNeighborsClassifier(n_neighbors=k)
    knn_classifier_scores=cross_val_score(knnclassifier, X, y, cv=10)
    print(statistics.mean(knn_classifier_scores)*100)

99.0
99.75
98.75
98.75
97.75
98.0
97.25
97.5
97.5
97.5
97.5
97.5
97.0
97.0
96.25
96.25
96.0
96.0
96.0
96.0
95.25
95.5
95.0
95.25
95.0
95.0
94.5
94.75
94.25
94.25
93.5
93.5
93.0
93.25
92.25
92.25
91.75
92.25
91.5
91.5
90.75
91.0
90.0
90.5
89.0
89.0
89.0
89.0
88.75


Accuracy of KNN is changing with change in K.Value of K makes a substantial difference in KNN model.Accuracy of the KNN is high when k=2.Accuracy of KNN is 99.75

# Naive Bayes

# MultinomialNB

In [200]:
from sklearn.naive_bayes import MultinomialNB

In [201]:
MultinomialNB_classifier =MultinomialNB()
MultinomialNB_classifier_scores=cross_val_score(MultinomialNB_classifier, X, y, cv=10)
statistics.mean(MultinomialNB_classifier_scores)*100

93.25

Accuracy of MultinomialNB with 10 fold cross validation is 93.25

# GaussianNB

In [203]:
from sklearn.naive_bayes import GaussianNB

In [204]:
GaussianNB_classifier =GaussianNB()
GaussianNB_classifier_scores=cross_val_score(GaussianNB_classifier, X, y, cv=10)
statistics.mean(GaussianNB_classifier_scores)*100

97.5

Accuracy of GaussianNB with 10 fold cross validation is 97.5.Accuracy of GaussianNB is higher than MultinomialNB

Conclusion: All the models are classifying well. Among all the models GaussianNB is having the highest accuracy and multinomial NB is having the lowest accuracy.